src/zoo.js

```diff
@@ -1,79 +1,241 @@
 };
 
 function schedule(dayName) {
-  // seu código aqui
+  const days = dayName ? [dayName] : Object.keys(data.hours);
+  return days.reduce((acc, day) => {
+    const { open, close } = data.hours[day];
+    const text = day === 'Monday'
+      ? 'CLOSED'
+      : `Open from ${open}am until ${close - 12}pm`
+    acc[day] = text;
+    return acc;
+  }, {});
 };
 
 function animalCount(species) {
-  // seu código aqui
+  if (species === undefined) {
+    const animaisCount = data.animals.reduce((emptyValue, CurrentValue) => {
+      emptyValue[`${CurrentValue.name}`] = CurrentValue.residents.length
+      return emptyValue
+    }, {})
+    return animaisCount
+  }
+
+  return data.animals.find(animal => animal.name === species).residents.length
 };
 
+function primeiroRequisito() {
+  return data.animals.reduce((emptyValue, CurrentValue) => {
+    if (emptyValue[CurrentValue.location] === undefined) {
+      emptyValue[CurrentValue.location] = [];
+    }
+    emptyValue[CurrentValue.location] =
+    [...emptyValue[CurrentValue.location], CurrentValue.name];
+    return emptyValue;
+  }, {})
+}
+
+function segundoRequisito() {
+  return data.animals.reduce((emptyValue, CurrentValue) => {
+    if (emptyValue[CurrentValue.location] === undefined) {
+      emptyValue[CurrentValue.location] = [];
+    }
+    emptyValue[CurrentValue.location] =
+    [...emptyValue[CurrentValue.location], { [CurrentValue.name]: CurrentValue.residents
+      .map(resident => resident.name) }]
+    return emptyValue;
+  }, {})
+}
+
+function terceiroRequisito() {
+  return data.animals.reduce((emptyValue, CurrentValue) => {
+    if (emptyValue[CurrentValue.location] === undefined) {
+      emptyValue[CurrentValue.location] = [];
+    }
+    emptyValue[CurrentValue.location] =
+    [...emptyValue[CurrentValue.location], { [CurrentValue.name]: CurrentValue.residents
+      .map(resident => resident.name).sort() }]
+    return emptyValue;
+  }, {})
+}
+
+function quartoRequisito() {
+  return data.animals.reduce((emptyValue, CurrentValue) => {
+    if (emptyValue[CurrentValue.location] === undefined) {
+      emptyValue[CurrentValue.location] = [];
+    }
+    emptyValue[CurrentValue.location] =
+    [...emptyValue[CurrentValue.location], { [CurrentValue.name]: CurrentValue.residents
+      .filter(resident => resident.sex === 'female').map(animal => animal.name) }]
+    return emptyValue;
+  }, {})
+}
+
 function animalMap(options) {
-  // seu código aqui
+  if (!options || (options.sex && !options.includeNames)) {
+    return primeiroRequisito()
+  }
+  if (options.sorted) {
+    return terceiroRequisito()
+  }
+  if (options.sex) {
+    return quartoRequisito()
+  }
+  return segundoRequisito()
 };
 
 function animalPopularity(rating) {
   // seu código aqui
 };
 
-function animalsByIds(ids) {
-  // seu código aqui
+function animalsByIds(...ids) {
+  const expected = [];
+  if (ids === undefined) {
+    return expected;
```

Comment:
Como aqui você não vai utilizar mais o expected para nenhum tipo de push, poderia retornar o array vazio direto.
Já te economizaria uma linha ;)
```suggestion
    return [];
```

Reply:
Humm, entendi a idéia, valeu